### PR TITLE
Make host names configurable

### PR DESCRIPTION
**Pull request by mamurak**
Pull request for issue #74 

Main changes:
- Host name of `pfioh` can be configured via environment variable `PFIOH_HOST_NAME`. If not set, the old value `pfioh_service` is used.
- Host name of `pman` can be configured via environment variable `PMAN_HOST_NAME`. If not set, the old value `pman_service` is used.

Additional changes:
- Extracted requirements from `setup.py` to dedicated requirements file (`requirements.txt`) to speed up container build. Pinned package versions.
- Fixed `pfurl` references in tests.
- Added `.dockerignore` file.
- Added tests to container build.
- Replaced base image by official Python image.
- Refactored and cleaned up Dockerfile and `ThreadedHTTPServer` methods.

### CODE REVIEW

**Comment by mamurak** [comment]
Investigating failed tests.

**Comment by jennydaman** [comment]
I suspect that the failed tests on Travis are not your fault, but something wrong with Travis (and our weak versioning).

https://github.com/FNNDSC/ChRIS_ultron_backEnd/pull/273

In this PR all I did was change documentation and the build fails for the same reason.


```
l_status = d_response['jobOperationSummary']['compute']['return']['l_status']
TypeError: string indices must be integers
```

**Comment by rudolphpienaar** [comment]
Have you tested these changes locally? Do you see the same failures?

**Comment by mamurak** [comment]
The CI agent can't resolve the `pfcon` dependencies, which is due to my adaptations. My local tests have passed, though.
I suspect the CI may need to be adapted since other PRs are failing as well. I'll look into it.

**Comment by jennydaman** [comment]
@rudolphpienaar `./make.sh` locally succeeds

**Comment by junaruga** [comment]
@mamurak Thank you for the work.
I love the pined pypi package version.
What is the benefits for us to have the host names as an environment variables?

File: Dockerfile
Comment (junaruga) [inline Dockerfile:28]: Is this 2 spaces indent for `Dockerfile` the best practice or common habit in FNNDSC? I was not familiar with it.

File: pfcon/pfcon.py
Comment (junaruga) [inline pfcon/pfcon.py:2027]: I see the IP address `8.8.8.8` for DNS is hardcorded. Is it fine not to be configurable?

**Comment by mamurak** [comment]
The CI agent can't resolve the `pfcon` dependencies, which is due to my adaptations. My local tests have passed, though.
I suspect the CI may need to be adapted since other PRs are failing as well. I'll look into it. 

> @mamurak Thank you for the work.
> I love the pined pypi package version.
> What is the benefits for us to have the host names as an environment variables? (portability and maintainability?)

The host names are currently hard coded. For running the service in a k8s environment, these host names have to be adapted as they're currently incompatible with the k8s naming scheme. Environment variables are the least invasive method for providing environment specific information such as third party host names such that we don't need to change the source code if the environment changes.

File: Dockerfile
Comment (mamurak) [reply Dockerfile:28]: It was best practice in my previous team. I think it greatly helps to visualize the different build stages.

Comment (mamurak) [reply Dockerfile:28]: Now I'm seeing a rogue comment in line 26, which should be removed. Thanks for your hint, @junaruga :)

File: pfcon/pfcon.py
Comment (mamurak) [reply pfcon/pfcon.py:2027]: That was part of the original code. I just reformatted it for clarity. To be honest, I don't know what this piece of code is doing exactly...

Comment (jennydaman) [reply pfcon/pfcon.py:2027]: It's a contrived way to get the machine's local IP address, that's it. @rudolphpienaar probably put it in for convenience.

https://github.com/FNNDSC/pfcon/pull/76/files/065189c5531c5e530b37d39c0a62f38c5f1a2d9c#diff-bedea3f9ced94f40f2c9b6cf2cd7c9ed25d3f515036a3f3d2817ffe0600f6c41L1882

The line is found repeated in many of our repos:

- https://github.com/FNNDSC/pfurl/blob/f355ef18d59b153d0ef03915b02ba241a17aefac/bin/pfurl#L23
- https://github.com/FNNDSC/pman/blob/d65f9aa500924c7c1fb7cb965c546fd000ae4c87/bin/pman#L31

IMO the entire functionality of reading the private IP address is bloat. If we want to keep it for convenience, there are shorter ways of figuring it out. Though less adaptable -- that socket connection to 8.8.8.8 is a fallback workaround in case a previous try to find out the IP did not work in an esoteric environment. See https://stackoverflow.com/questions/166506/finding-local-ip-addresses-using-pythons-stdlib/166589#166589

I was scratching my head about this a few months ago too, if you want a thorough explanation of what's happening please see [pfmisc/local_ip.py](https://github.com/FNNDSC/pfmisc/blob/master/pfmisc/local_ip.py) where I flattened the control flow and commented.

Comment (mamurak) [reply pfcon/pfcon.py:2027]: Great, thanks for the insights @jennydaman :+1:

File: Dockerfile
Comment (junaruga) [reply Dockerfile:28]: Okay. Thanks for explaining it!

File: pfcon/pfcon.py
Comment (junaruga) [reply pfcon/pfcon.py:2027]: > That was part of the original code.

Okay. @mamurak @jennydaman thanks for sharing the context and documents!
It seems the 8.8.8.8 is Google's public DNS server IP.
https://developers.google.com/speed/public-dns/

**Comment by junaruga** [comment]
@mamurak All right! Thanks for clarifying the benefits!

File: pfcon/pfcon.py
Comment (rudolphpienaar) [reply pfcon/pfcon.py:2027]: Yes, I am the source of these convolutions. The reason it was necessary arose the proxied env of our Institution network. The short story is that addresses such as `localhost` were not handled consistently in the proxied env and hence I needed to code/borrow some quick method to find the actual host IP address (and not `localhost` or `127.0.0.1`).

Yes, 8.8.8.8 is one of google's public DNS -- it has been that way for many years so I assume that IP is reliable/stable.

**Comment by mamurak** [comment]
Removed the `pfcon` integration tests as they're already executed during container build. They've been failing as the CI agent didn't find the dependencies, which I had moved to a dedicated requirements file.

**Comment by mamurak** [comment]
Rebased feature branch onto latest master commits.

**Comment by jennydaman** [comment]
hey, I fixed CUBE (it was my fault).

Integration tests should pass now.

File: Dockerfile
Comment (jennydaman) [inline Dockerfile:55]: `RUN rm -fr ${APPROOT}` is an effort to minimize the published image size, so it should be part of the `runtime` stage. Though it might be futile because it's adding a layer.

File: Dockerfile
Comment (jennydaman) [inline Dockerfile:27]: While we're at it, it'd be good to modernize `Dockerfile`.

`MAINTAINER` is deprecated, use

```Dockerfile
LABEL version="3.0.0.0" maintainer="FNNDSC <dev@babyMRI.org>"
```

File: pfcon/pfcon.py
Comment (jennydaman) [inline pfcon/pfcon.py:2001]: f-strings are new in python3.6, so `setup.py` also needs modernizing.

example: https://github.com/FNNDSC/pfioh/blob/master/setup.py

**Review by jennydaman — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (mamurak) [reply Dockerfile:55]: Good catch. Yes, I should have moved it to the `runtime` stage. But as you noted, we're not gaining much in terms of image size; that step takes a noticeable amount of time during building, however. So I'm going to remove it.

To optimize the image size, we may want to adjust the CI and Dockerfile as indicated [here](https://pythonspeed.com/articles/faster-multi-stage-builds/). I'll leave this to a future task as that matter isn't as pressing, though :)

File: pfcon/pfcon.py
Comment (mamurak) [reply pfcon/pfcon.py:2001]: Github is indicating this conversation as unresolved. Could you please double-check whether this needs more work, @jennydaman?

File: .github/workflows/ci.yml
Comment (jennydaman) [inline .github/workflows/ci.yml:13]: I think we've settled on (but without fully agreeing) that we should keep unit tests outside `Dockerfile` and `docker build`. Would you revert that change?

p.s. multi-stage builds are nice but if we don't need it then we should not use it -- one of our Power9 machines on the MOC is running RHEL 7.6 with Docker version 1.13.1, which does not support multi-stage builds.

p.p.s. so to take advantage of fancy build features, images must be built on some other machine first (really, on my laptop using QEMU emulation) then exported to that machine. Cross-compilation in CI is something we can explore as future steps, but this is all very off-topic.

**Review by jennydaman — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml
Comment (mamurak) [reply .github/workflows/ci.yml:13]: I removed the test stage.

File: Dockerfile
Comment (jennydaman) [inline Dockerfile:26]: There is no longer a need for `AS build`, see https://github.com/FNNDSC/pfcon/pull/76/files#r550116964

File: .github/workflows/ci.yml
Comment (jennydaman) [reply .github/workflows/ci.yml:13]: Could you restore the nose tests back to `ci.yml` since it has been removed from `Dockerfile`?

**Review by jennydaman — CHANGES_REQUESTED** [review]

**Review by jennydaman — APPROVED** [review]

**Review by jennydaman — COMMENTED** [review]
c2889e1a72d7ba40f100a339c8de644b7d7cbf70

My understanding is that `requirements.txt` and `install_requires` in `setup.py` are not an either/or, they represent different things:

- `install_requires` is an abstract description of the dependencies
- `requirements.txt` is a precise description of the dependencies in its last working state

Nonetheless there is no convergence on practices within the broader Python community so I am not sure what to think about it.

**Comment by mamurak** [comment]
> [c2889e1](https://github.com/FNNDSC/pfcon/commit/c2889e1a72d7ba40f100a339c8de644b7d7cbf70)
> 
> My understanding is that `requirements.txt` and `install_requires` in `setup.py` are not an either/or, they represent different things:
> 
>     * `install_requires` is an abstract description of the dependencies
> 
>     * `requirements.txt` is a precise description of the dependencies in its last working state
> 
> 
> Nonetheless there is no convergence on practices within the broader Python community so I am not sure what to think about it.

Yes, you're right, @jennydaman. I'd consider the following to be a good practice:
* `install_requires` defines the range of compatible version, i.e. the lowest compatible and the highest one if it exists.
* `requirements.txt` defines a set of specific versions to be installed for reproducibility.

The first point is particularly important in case of libraries that we're distributing throughout different projects. The second point (pinned versions) is particularly important for applications, which need consistent dependencies across development and production stages. But also while developing libraries, pinning package version is useful.
Check out [this article|https://caremad.io/posts/2013/07/setup-vs-requirement/] for a perspective on these two use cases.

**Comment by jennydaman** [comment]
So do we want to restore the `install_requires` line before merging?

**Comment by mamurak** [comment]
> So do we want to restore the `install_requires` line before merging?

Since `pfcon` is not a library, I'm fine with leaving out `install_requires`.

**Pull request merged**
